runtime/lua/vim/treesitter/query_linter.lua

```diff
@@ -0,0 +1,153 @@
+local namespace = vim.api.nvim_create_namespace('vim.treesitter.query_linter')
+-- those node names exist for every language
+local BUILT_IN_NODE_NAMES = { '_', 'ERROR' }
+
+local M = {}
+
+---@private
+--- Adds a diagnostic for node in the query buffer
+local function add_lint_for_node(diagnostics, node, buf, lint, lang)
+  local node_text = vim.treesitter.get_node_text(node, buf):gsub('\n', ' ')
+  local message = lint .. ': ' .. node_text
+  local error_range = { node:range() }
+  table.insert(diagnostics, {
+    lnum = error_range[1],
+    end_lnum = error_range[3],
+    col = error_range[2],
+    end_col = error_range[4],
+    severity = vim.diagnostic.ERROR,
+    message = message,
+    source = lang,
+  })
+end
+
+---@private
+--- like vim.tbl_contains but works with function predicate
+local function table_contains(predicate, table)
+  for _, elt in pairs(table) do
+    if predicate(elt) then
+      return true
+    end
+  end
+  return false
+end
+
+---@private
+--- Determines the target language of a query file by its path: <lang>/<query_type>.scm
+local function guess_query_lang(buf, _)
+  local filename = vim.api.nvim_buf_get_name(buf)
+  if filename ~= '' then
+    local ok, query_lang = pcall(vim.fn.fnamemodify, filename, ':p:h:t')
+    return ok and query_lang
+  end
+end
+
+---@private
+local function normalize_opts(buf, opts)
+  opts = opts or {}
+  opts.langs = opts.langs or guess_query_lang
+  if type(opts.langs) == 'function' then
+    opts.langs = opts.langs(buf, opts)
+  end
+  if type(opts.langs) ~= 'table' then
+    opts.langs = { opts.langs }
+  end
+  return opts
+end
+
+local lint_query = [[
+  (program [(named_node) (list) (grouping)] @toplevel)
+  (named_node
+    name: _ @node.named)
+  (anonymous_node
+    name: _ @node.anonymous)
+  (field_definition
+    name: (identifier) @field)
+  (ERROR) @error
+]]
+
+--- Determines diagnostics for a buffer containing tree-sitter queries (|treesitter-query|).
+--- For runtime query files that target installed tree-sitter parsers, this function will use
+--- tree-sitter API to determine admissible identifiers to describe nodes of those parsers and
+--- verifies that top-level s-expressions can be loaded in isolation by Neovim to debug errors
+--- in runtime query files. The found diagnostics are reported using |diagnostic-api|.
+--- @param buf integer Buffer to lint
+--- @param opts nil|table Options for linting
+--- - langs nil|string|string[]|function One or more parsers which the current queries target (see
+--        |treesitter-parsers) or a function that determines the list of parsers given `buf`,
```

Comment:
```suggestion
--        |treesitter-parsers|) or a function that determines the list of parsers given `buf`,
```

Reply:
Is there a possibilty to have module level documentation? I would like to have "what is this?" module level" and only "how to use it?" as documentation for this function.

Reply:
Yes, with `@defgroup` -- look at `runtime/lua/vim/treesitter/languagetree.lua`.

Reply:
Moved this into `vim.treesitter` to behave like `playground`. Maybe `vim.treesitter.devtools` in future?

Reply:
Nope; that was roundly rejected.

Reply:
So, implementation file with or without underscore? `_fold.lua` has underscore, `playground` not. Would prefer with underscore